File: scripts/normalize_url.py
#!/usr/bin/env python3
"""Normalize a URL for deduplication.

Applies the canonical normalization rules from FILTER.md Rule 0:
strip tracking params, fragments, trailing slashes, AMP paths,
and normalize scheme/host.

Usage:
    python3 scripts/normalize_url.py <url>
    echo "<url>" | python3 scripts/normalize_url.py

Output: normalized URL to stdout (one line).
"""

import sys
import re
from urllib.parse import urlparse, urlunparse, parse_qs, urlencode

# Query parameters to strip (tracking, attribution, session)
STRIP_PARAMS = {
    "utm_source", "utm_medium", "utm_campaign", "utm_term", "utm_content",
    "fbclid", "gclid", "gclsrc", "ref", "source", "_ga", "_gid",
    "ncid", "ocid", "sr_share", "mc_cid", "mc_eid",
}


def normalize_url(url):
    """Normalize a URL for deduplication."""
    url = url.strip()
    if not url:
        return url

    # Ensure scheme
    if not url.startswith(("http://", "https://")):
        url = "https://" + url

    parsed = urlparse(url)

    # Normalize scheme: http -> https
    scheme = "https"

    # Normalize host: lowercase, strip www., strip amp. subdomain
    host = parsed.hostname or ""
    host = host.lower()
    if host.startswith("www."):
        host = host[4:]
    if host.startswith("amp."):
        host = host[4:]

    # Preserve port if non-standard
    port = parsed.port
    netloc = f"{host}:{port}" if port and port not in (80, 443) else host

    # Normalize path: strip AMP suffixes, trailing slashes
    path = parsed.path
    path = re.sub(r"/amp/?$", "", path)          # /amp or /amp/
    path = re.sub(r"\.amp\.html$", ".html", path) # .amp.html -> .html
    path = path.rstrip("/") or "/"

    # Strip tracking query params
    if parsed.query:
        params = parse_qs(parsed.query, keep_blank_values=True)
        filtered = {k: v for k, v in params.items() if k.lower() not in STRIP_PARAMS}
        query = urlencode(filtered, doseq=True) if filtered else ""
    else:
        query = ""

    # Strip fragment entirely
    fragment = ""

    return urlunparse((scheme, netloc, path, "", query, fragment))


def main():
    if len(sys.argv) > 1:
        url = sys.argv[1]
    else:
        url = sys.stdin.readline()

    result = normalize_url(url)
    if result:
        print(result)


if __name__ == "__main__":
    main()

File: scripts/normalize-url.ts
/**
 * Normalize a URL for deduplication.
 *
 * Strips tracking params, fragments, trailing slashes, AMP paths,
 * and normalizes scheme/host.
 *
 * Usage:
 *   npx tsx scripts/normalize-url.ts <url>
 *   echo "<url>" | npx tsx scripts/normalize-url.ts
 *
 * Importable:
 *   import { normalizeUrl } from './normalize-url.js';
 */

const STRIP_PARAMS = new Set([
  "utm_source", "utm_medium", "utm_campaign", "utm_term", "utm_content",
  "fbclid", "gclid", "gclsrc", "ref", "source", "_ga", "_gid",
  "ncid", "ocid", "sr_share", "mc_cid", "mc_eid",
]);

export function normalizeUrl(raw: string): string {
  let url = raw.trim();
  if (!url) return url;

  // Ensure scheme
  if (!url.startsWith("http://") && !url.startsWith("https://")) {
    url = "https://" + url;
  }

  const parsed = new URL(url);

  // Normalize scheme
  parsed.protocol = "https:";

  // Normalize host: lowercase, strip www. and amp. subdomains
  let host = parsed.hostname.toLowerCase();
  if (host.startsWith("www.")) host = host.slice(4);
  if (host.startsWith("amp.")) host = host.slice(4);
  parsed.hostname = host;

  // Strip non-standard port (URL class handles 80/443 automatically)
  // No action needed — new URL() already omits default ports

  // Normalize path: strip AMP suffixes, trailing slashes
  let path = parsed.pathname;
  path = path.replace(/\/amp\/?$/, "");
  path = path.replace(/\.amp\.html$/, ".html");
  path = path.replace(/\/+$/, "") || "/";
  parsed.pathname = path;

  // Strip tracking query params
  const params = parsed.searchParams;
  const keysToDelete: string[] = [];
  for (const key of params.keys()) {
    if (STRIP_PARAMS.has(key.toLowerCase())) {
      keysToDelete.push(key);
    }
  }
  for (const key of keysToDelete) {
    params.delete(key);
  }

  // Strip fragment
  parsed.hash = "";

  return parsed.toString();
}

// --- CLI entry point ---

async function main() {
  let url: string;

  if (process.argv.length > 2) {
    url = process.argv[2];
  } else {
    // Read from stdin
    const chunks: Buffer[] = [];
    for await (const chunk of process.stdin) {
      chunks.push(chunk as Buffer);
    }
    url = Buffer.concat(chunks).toString("utf-8").trim();
  }

  const result = normalizeUrl(url);
  if (result) {
    console.log(result);
  }
}

// Run CLI if invoked directly
const isMain = process.argv[1]?.endsWith("normalize-url.ts") ||
               process.argv[1]?.endsWith("normalize-url.js");
if (isMain) {
  main();
}
